llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp

```diff
@@ -182,26 +184,40 @@ StackSlot ControlFlowGraphBuilder::getDefiningSlot(const MachineInstr &MI,
 }
 
 void ControlFlowGraphBuilder::collectInstrOperands(const MachineInstr &MI,
-                                                   Stack &Input,
-                                                   Stack &Output) const {
-  for (const auto &MO : reverse(MI.explicit_uses())) {
-    if (!MO.isReg()) {
-      if (MO.isMCSymbol())
-        Input.push_back(SymbolSlot{MO.getMCSymbol()});
-      continue;
-    }
+                                                   Stack *Input,
+                                                   Stack *Output) const {
+  if (Input) {
+    for (const auto &MO : reverse(MI.explicit_uses())) {
+      if (!MO.isReg()) {
+        if (MO.isMCSymbol())
+          Input->push_back(SymbolSlot{MO.getMCSymbol()});
+        continue;
+      }
 
-    const Register Reg = MO.getReg();
-    // SP is not used anyhow.
-    if (Reg == EVM::SP)
-      continue;
+      const Register Reg = MO.getReg();
+      // SP is not used anyhow.
+      if (Reg == EVM::SP)
+        continue;
 
-    Input.push_back(getDefiningSlot(MI, Reg));
+      Input->push_back(getDefiningSlot(MI, Reg));
+    }
   }
 
-  unsigned ArgsNumber = 0;
-  for (const auto &MO : MI.defs())
-    Output.push_back(TemporarySlot{&MI, MO.getReg(), ArgsNumber++});
+  if (Output) {
+    unsigned ArgsNumber = 0;
+    for (const auto &MO : MI.defs())
+      Output->push_back(TemporarySlot{&MI, MO.getReg(), ArgsNumber++});
+  }
+}
+
+bool ControlFlowGraphBuilder::isCommutable(const MachineInstr &MI) const {
+  // In all the cases, commutable operands take top two stack slots.
+  // Even for ADDMOD and MULMOD instructions, because their MOD operand
+  // takes third slot counted from the top of stack.
+  unsigned Opc = MI.getOpcode();
+  return (MI.isCommutable() || Opc == EVM::ADDMOD || Opc == EVM::MULMOD)
+             ? true
```

Comment:
Redundant ternary

Reply:
Yep, thank you!

Reply:
Sorry for jumping in, but is there any reason why we didn't mark `ADDMOD` and `MULMOD` as commutable in `EVMInstrInfo.td`?

Reply:
Thank you for pointing this out. I didn't put `isCommutable` on these instructions, because of non-commutable third argument (module value). I've just I looked at the MachineInstr::isCommutable() internals and other related functions and it looks like we can define them with `isCommutable` flag, because a commutable instruction may have form  `X = op Y, Z, ...`, where only Y and Z operands are commutable.

Reply:
Then indeed it's better to use the flag.